File: src/main/java/com/nikitavbv/assignments/discretestructures/lab1/stack/MyFixedStack.java
package com.nikitavbv.assignments.discretestructures.lab1.stack;

import java.util.Arrays;
import java.util.NoSuchElementException;

/**
 * Stack implementation with fixed capacity.
 *
 * @param <T> the type of elements in this stack
 *
 * @author Nikita Volobuev
 * @see MyStackInterface
 */
public class MyFixedStack<T> implements MyStackInterface<T> {

    private static final int DEFAULT_CAPACITY = 100;

    T[] arr;
    private int pos = 0;

    /** Creates stack with default capacity */
    public MyFixedStack() {
        this(DEFAULT_CAPACITY);
    }

    /**
     * Creates stack with specified capacity.
     *
     * @param capacity maximum size of this stack.
     * @throws IllegalArgumentException if stack capacity is < 0
     */
    public MyFixedStack(int capacity) {
        if (capacity < 0) {
            throw new IllegalArgumentException("Capacity should be >= 0");
        }

        // noinspection unchecked
        this.arr = (T[]) new Object[capacity];
    }

    @Override
    public void push(T element) {
        if (pos >= arr.length) {
            throw new StackFullException();
        }

        arr[pos] = element;
        pos++;
    }

    @Override
    public T pop() {
        if (isEmpty()) {
            throw new NoSuchElementException();
        }

        pos--;
        final T element = arr[pos];
        arr[pos] = null;
        return element;
    }

    @Override
    public T top() {
        if (isEmpty()) {
            throw new NoSuchElementException();
        }

        return arr[pos - 1];
    }

    @Override
    public void clear() {
        pos = 0;
        Arrays.fill(arr, null);
    }

    @Override
    public boolean isEmpty() {
        return getSize() == 0;
    }

    @Override
    public int getSize() {
        return pos;
    }

    public int getCapacity() {
        return arr.length;
    }
}

File: src/main/rust/discreteStructures/stack/src/my_fixed_stack.rs
use crate::my_stack_trait::MyStack;

pub struct MyFixedStack<T: Copy> {
    arr: Box<[Option<T>]>,
    pos: usize,
    capacity: usize,
}

impl <T: Copy> MyFixedStack<T> {
    const DEFAULT_CAPACITY: usize = 100;

    pub fn new(capacity: usize) -> MyFixedStack<T> {
        let mut vec = Vec::with_capacity(capacity);
        for _i in 0..capacity {
            vec.push(None);
        }
        MyFixedStack {
            arr: vec.into_boxed_slice(),
            pos: 0,
            capacity
        }
    }

    pub fn with_default_capacity() -> MyFixedStack<T> {
        Self::new(Self::DEFAULT_CAPACITY)
    }
}

impl <T: Copy> MyStack<T> for MyFixedStack<T> {
    fn push(&mut self, element: T) {
        if self.pos >= self.capacity {
            panic!("Stack is full")
        }

        self.arr[self.pos] = Some(element);
        self.pos += 1;
    }

    fn pop(&mut self) -> Option<T> {
        if self.is_empty() {
            return None
        }

        self.pos -= 1;
        self.arr[self.pos]
    }

    fn top(&mut self) -> Option<T> {
        if self.is_empty() {
            return None
        }

        self.arr[self.pos - 1]
    }

    fn clear(&mut self) {
        self.pos = 0;
    }

    fn is_empty(&self) -> bool {
        self.get_size() == 0
    }

    fn get_size(&self) -> usize {
        self.pos
    }
}